src/app/logic/upload.ts

```diff
@@ -0,0 +1,74 @@
+import worker from "../spawn-worker";
+
+/**
+ * Returns an array of all files uploaded from a drag-and-drop event.
+ */
+export function filesFromDragEvent(event: DragEvent): File[] {
+  if (!event.dataTransfer) {
+    return [];
+  } else if (event.dataTransfer.items) {
+    // Use DataTransferItemList interface to access the file(s)
+    return filesFromDataTransferItems(event.dataTransfer.items);
+  } else {
+    // Use DataTransfer interface to access the file(s)
+    return Array.from(event.dataTransfer.files);
+  }
+}
+
+/**
+ * Returns an array of files from the given event target, which is assumed to
+ * be an <input type="file">.
+ */
+export function filesFromInputElement(el: EventTarget | null): File[] {
+  if (el === null) return [];
+  if (!isFileInputElement(el)) return [];
```

Comment:
I have complicated thoughts on this!

In general, my answer is yes, errors are good. In particular, this _needs_ to be called on a `<input type="file">` element, so both `if (el === null)` and `if (!FileInputElement())` should raise some sort of assertion error (a.k.a., programmer's assumptions are invalid). However, who would catch that error? This code is destined for the app bundle, so logging via `console.error()` would only serve to clutter up users' console logs, that most users won't see. If we were using some crash reporting (e.g., Sentry or Rollbar), we would send that error here.

Haha, so with that 4 dimensional chess out of the way, I decided just to forget about raising error messages, because errors in browser code are annoying :/

Reply:
That explanation makes sense to me and is useful for gaining insight into front-end development, thanks :+1:!

If you don't mind, I'm curious about my current assumptions about using the browser console. My current rationale in favour of using the console is exactly because most users can't see it, so that developers can take a peek when intermediate processes have gone awry to save time debugging. I agree that it's never good when the console is filled with needless logs, but do you think using `console.error()` sparingly is acceptable? I'm curious about this subject as it's one I don't have a lot of experience in.

As this project expands in complexity in its logic, do you think we should add a ticket to add something like Sentry or Rollbar? If not crash reporting or logging, is there some other solution for flagging intermediate errors to help developers pinpoint bugs?

Once again, I appreciate picking your brain on these subjects!

Reply:
> I agree that it's never good when the console is filled with needless logs, but do you think using console.error() sparingly is acceptable? I

I don't have a clear answer for this, unfortunately. My biggest issue against it is that [`console.error()` is a memory leak](https://cherniavskii.com/be-careful-with-console-logs/#console-warnings). But yes, using it sparingly — if you can guarentee it once per page load — is acceptable.

> As this project expands in complexity in its logic, do you think we should add a ticket to add something like Sentry or Rollbar? If not crash reporting or logging, is there some other solution for flagging intermediate errors to help developers pinpoint bugs?

Sure! I'm not sure about other solutions for flagging intermediate errrors. There are static analysis tools, including the TypeScript typechecker itself. But you have to write clever code in TypeScript to be able to take advantage of it.